File: agent/streaming/events.py
# 流式事件：Agent 运行过程中向 EventSink 推送的事件类型。
#
# 三层（对应 claude-code 的双层事件模型，2026-08-08 对齐 CC 源码）：
# - 低层（adapter 发）：TextDelta / ToolCallStart / ToolCallDelta / ToolCallEnd / MessageEnd
#   —— LLM 流式输出增量（逐 token 文本、工具参数 JSON 增量）。仅 CLI 打字机消费。
# - 高层（loop / executor 发）：RunStart / StepStart / ToolStart / ToolEnd / StepEnd / RunEnd
#   —— Agent 生命周期与工具执行进度。
# - 消息级（loop 源头发）：AssistantMessage / ToolResultMessage
#   —— 完整 assistant 消息与工具结果，数据在 ModelResponse/ToolResult 处已齐，
#      直接发整包（对齐 CC QueryEngine 的 `assistant`/`user(tool_result)`，不做 delta 聚合）。
#      这是 web/SSE 消费的契约（server.py _is_web_event 白名单）；delta 只在 CLI 打字机用。
#
# 设计：frozen dataclass + StreamEvent Union；sink 用 match(event) 分发。
# 事件层不依赖 state / models / adapters，保持叶子层（与 enums 同级）。
from dataclasses import dataclass
from typing import Any, Union


# ─────────────────── 消息级：完整内容（loop 源头发，web 契约） ───────────────────

@dataclass(frozen=True)
class AssistantMessage:
    """一条完整的 assistant 消息（对齐 CC `assistant`）。

    _run_steps 拿到完整 ModelResponse（text/thinking/tool_calls/usage 全量）时源头发，
    不做 delta 聚合。tool_calls 是已解析 dict 列表 {call_id, tool_name, arguments}。
    uuid 全程带（前端去重/关联，对齐 CC）；agent_id 打标子 agent（多 Agent 展示用）。
    """
    run_id: str
    uuid: str
    agent_id: str | None = None
    step_index: int | None = None
    text: str = ""
    thinking: str = ""
    tool_calls: tuple = ()      # tuple[dict]: {call_id, tool_name, arguments}
    stop_reason: str | None = None
    usage: Any = None           # TokenUsage；用 Any 避免事件层反向依赖 models


@dataclass(frozen=True)
class ToolResultMessage:
    """一条完整工具结果（对齐 CC `user` 的 tool_result 块）。

    工具结果就绪处（_run_steps/_run_workflow/_execute_pending 的 async-for）源头发。
    elapsed_ms 由 ToolResult.elapsed_ms 传入（_parallel.run_one 实测，修待办 C 恒 0）。
    """
    run_id: str
    uuid: str
    call_id: str
    tool_name: str
    ok: bool
    summary: str | None = None
    elapsed_ms: float = 0.0
    attempts: int = 1
    error_type: str | None = None
    agent_id: str | None = None


# ─────────────────── 高层：生命周期（loop / executor 发） ───────────────────

@dataclass(frozen=True)
class RunStart:
    """一次 agent 运行开始"""
    run_id: str


@dataclass(frozen=True)
class StepStart:
    """一轮 Agent 循环开始"""
    step_index: int


@dataclass(frozen=True)
class StepEnd:
    """一轮 Agent 循环结束"""
    step_index: int


@dataclass(frozen=True)
class RunEnd:
    """一次 agent 运行结束（终态）。usage/duration_ms/num_steps = 本轮聚合统计
    （对齐 CC `result` 事件：整轮所有 assistant 消息 usage 累加，turn 结束才报总账；
    与每条 assistant 消息自带的 per-step usage 互补）。"""
    status: str
    final_text: str | None = None
    error: dict[str, Any] | None = None
    usage: dict[str, int] | None = None      # 本轮聚合:input/output/total/cached(不含子 agent,子 agent 各自累计)
    duration_ms: float | None = None         # 本轮耗时(首 step start ~ 末 step end)
    num_steps: int | None = None             # 本轮 step 数(对齐 CC result.num_turns)


@dataclass(frozen=True)
class ToolStart:
    """即将执行某工具（executor 发）"""
    call_id: str
    tool_name: str
    arguments: dict[str, Any]


@dataclass(frozen=True)
class ToolEnd:
    """工具执行完毕（executor 发）"""
    call_id: str
    tool_name: str
    ok: bool
    elapsed_ms: float = 0.0
    error_type: str | None = None
    summary: str | None = None
    attempts: int = 1   # 实际尝试次数(含首次);>1 表示发生过重试,MetricsCollector 算 retry_count 用(#6)


@dataclass(frozen=True)
class ApprovalRequestEvent:
    """HITL:需人工批准的工具请求(web_confirmer 经 SSE 队列推前端弹窗)。

    不进 EventSink(web_confirmer 直接 put 到 SSE 队列,不走 sink/ tracer 链)。
    前端收到 -> 弹窗 -> POST /approve/{request_id} 解 future(见 guardrails/confirmer.py)。
    """
    request_id: str
    tool_name: str
    reason: str
    arguments: dict[str, Any]


# ─────────────────── 低层：LLM 流式增量（adapter 发） ───────────────────

@dataclass(frozen=True)
class TextDelta:
    """一段流式文本（可能几个字，也可能整段——非流式 fallback 时是整段）"""
    text: str


@dataclass(frozen=True)
class ThinkingDelta:
    """一段流式 thinking/reasoning 增量（模型内部 CoT，区别于最终回答 TextDelta）。
    DeepSeek 的 reasoning_content / ark 的 thinking / Claude 的 thinking block。
    expose_reasoning=False 时 printer 不渲染（对齐 CC 隐藏内部推理），最终回答(TextDelta)始终可见。"""
    text: str


@dataclass(frozen=True)
class ToolCallStart:
    """模型开始产出第 index 个工具调用（拿到 call_id / tool_name）"""
    call_id: str
    tool_name: str
    index: int = 0


@dataclass(frozen=True)
class ToolCallDelta:
    """工具调用参数 JSON 的增量片段"""
    call_id: str
    arguments_delta: str


@dataclass(frozen=True)
class ToolCallEnd:
    """一个工具调用的参数收完（可解析为完整 arguments）"""
    call_id: str


@dataclass(frozen=True)
class MessageEnd:
    """本轮 LLM 消息结束"""
    stop_reason: str | None = None
    usage: Any = None  # TokenUsage；用 Any 避免事件层反向依赖 models


StreamEvent = Union[
    # 消息级（web 契约）
    AssistantMessage, ToolResultMessage,
    # 高层
    RunStart, StepStart, StepEnd, RunEnd, ToolStart, ToolEnd, ApprovalRequestEvent,
    # 低层
    TextDelta, ThinkingDelta, ToolCallStart, ToolCallDelta, ToolCallEnd, MessageEnd,
]

File: chatweb/frontend/lib/events.ts
// lib/events.ts - web 消费的消息级事件(单点契约的 TS 镜像 + CC 外部 StructuredIO)
// 契约唯一真源:agent/streaming/events.py 的 is_web_event(WEB_EVENT_TYPES)。
// 后端 server.py SSE 过滤 与 EventStore 落盘(events.jsonl)共用同一判定;本文件必须与它一致:
//   RunStart / RunEnd          - turn 书签
//   AssistantMessage           - 完整 assistant 消息(每 step 一条,自带 usage/tool_calls)
//   ToolStart                  - 工具开始(resume/_workflow 无 LLM step 时给前端建卡)
//   ToolResultMessage          - 完整工具结果(call_id 关联,自带 elapsed_ms)
//   ApprovalRequestEvent       - HITL(ChatContext 拦截弹窗,不进消息体)
//   TaskNotification           - 后台子 agent 完成通知(自动 turn 开头推,渲染系统提示行)
// 关联键 call_id:AssistantMessage.tool_calls 建卡,ToolResultMessage 按 call_id 跨消息 patch。
// 若后端 events.jsonl 被用作历史恢复,reducer 可直接重放其中 {"type", ...} 记录(含 ts)。
import type { ChatMessage, ToolCallView, TokenUsage } from "./types";

export type StreamEvent =
  | { type: "RunStart"; run_id: string }
  | {
      type: "RunEnd";
      status: string;
      final_text: string | null;
      error: Record<string, unknown> | null;
      usage: TokenUsage | null; // 整轮聚合(对齐 CC result.usage;turn 结束才显示总账)
      duration_ms: number | null;
      num_steps: number | null;
    }
  | {
      type: "AssistantMessage";
      run_id: string;
      uuid: string;
      agent_id: string | null;
      step_index: number | null;
      text: string;
      thinking: string;
      tool_calls: { call_id: string; tool_name: string; arguments: Record<string, unknown> }[];
      stop_reason: string | null;
      usage: TokenUsage | null;
    }
  | { type: "ToolStart"; call_id: string; tool_name: string; arguments: Record<string, unknown> }
  | {
      type: "ToolResultMessage";
      run_id: string;
      uuid: string;
      call_id: string;
      tool_name: string;
      ok: boolean;
      summary: string | null;
      elapsed_ms: number;
      attempts: number;
      error_type: string | null;
      agent_id: string | null;
    }
  | { type: "ApprovalRequestEvent"; request_id: string; tool_name: string; reason: string; arguments: Record<string, unknown> }
  | {
      type: "TaskNotification"; // 后台子 agent 完成通知(自动 turn 开头推,渲染系统提示行)
      run_id: string;
      role: string;
      status: string;
      text: string;
    };

export interface ChatState {
  messages: ChatMessage[];
  streaming: boolean;
}

function toolCallView(tc: {
  call_id: string;
  tool_name: string;
  arguments: Record<string, unknown>;
}): ToolCallView {
  const args = tc.arguments ?? {};
  return {
    callId: tc.call_id,
    toolName: tc.tool_name,
    argumentsJson: JSON.stringify(args),
    arguments: args,
    phase: "producing",
  };
}

/** 按 call_id 找跨消息的 ToolCallView,返回 (messageIdx, toolIdx);找不到返回 null。 */
function findToolCall(msgs: ChatMessage[], callId: string): { m: number; t: number } | null {
  for (let m = 0; m < msgs.length; m++) {
    const tcs = msgs[m].toolCalls || [];
    for (let t = 0; t < tcs.length; t++) {
      if (tcs[t].callId === callId) return { m, t };
    }
  }
  return null;
}

function patchToolCall(
  msgs: ChatMessage[],
  callId: string,
  patch: Partial<ToolCallView>,
): ChatMessage[] {
  const found = findToolCall(msgs, callId);
  if (!found) return msgs;
  const { m, t } = found;
  return msgs.map((msg, i) =>
    i !== m
      ? msg
      : { ...msg, toolCalls: (msg.toolCalls || []).map((tc, j) => (j === t ? { ...tc, ...patch } : tc)) },
  );
}

export function initialState(): ChatState {
  return { messages: [], streaming: false };
}

/** 有状态 reducer:输入一个消息级事件,产出新的消息列表(每 step 独立气泡 + 按 call_id 关联)。 */
export function eventReducer(state: ChatState, ev: StreamEvent): ChatState {
  switch (ev.type) {
    case "RunStart":
      return { ...state, streaming: true };

    case "AssistantMessage": {
      // 幂等:uuid 已存在 -> 跳过(单通道下 restore 重放 events.jsonl 与 /stream 重连
      // 补发队列会重叠,同 uuid 的消息不重复 append)。
      if (state.messages.some((m) => m.id === ev.uuid)) return state;
      // 每 step 一条消息,自带 usage(修"只显示最后一步 token")
      const msg: ChatMessage = {
        id: ev.uuid,
        role: "assistant",
        content: ev.text,
        thinking: ev.thinking || undefined,
        toolCalls: ev.tool_calls.map(toolCallView),
        usage: ev.usage || undefined,
        streaming: true,
        createdAt: Date.now(),
      };
      return { ...state, messages: [...state.messages, msg], streaming: true };
    }

    case "ToolStart": {
      // 正常路径:AssistantMessage.tool_calls 已建卡,这里仅置 running。
      // resume/_workflow 无 LLM step:没有卡,用 ToolStart 建卡(参数来自 ToolStart.arguments)。
      const found = findToolCall(state.messages, ev.call_id);
      if (found) return { ...state, messages: patchToolCall(state.messages, ev.call_id, { phase: "running" }) };
      const tc: ToolCallView = {
        callId: ev.call_id,
        toolName: ev.tool_name,
        argumentsJson: JSON.stringify(ev.arguments ?? {}),
        arguments: ev.arguments ?? {},
        phase: "running",
      };
      const msgs = [...state.messages];
      const last = msgs[msgs.length - 1];
      if (last && last.role === "assistant") {
        return { ...state, messages: [...msgs.slice(0, -1), { ...last, toolCalls: [...(last.toolCalls || []), tc] }] };
      }
      return {
        ...state,
        messages: [...msgs, { id: "ph-" + ev.call_id, role: "assistant", content: "", toolCalls: [tc], streaming: true, createdAt: Date.now() }],
      };
    }

    case "ToolResultMessage": {
      const msgs = state.messages;
      if (!findToolCall(msgs, ev.call_id)) {
        // 兜底:结果到了但卡没了,append 一张已完成卡(不丢结果)
        const tc: ToolCallView = {
          callId: ev.call_id,
          toolName: ev.tool_name,
          argumentsJson: "{}",
          arguments: {},
          phase: ev.ok ? "done" : "error",
          ok: ev.ok,
          summary: ev.summary || undefined,
          elapsedMs: ev.elapsed_ms,
          attempts: ev.attempts,
          errorType: ev.ok ? undefined : ev.error_type || undefined,
          errorMessage: ev.ok ? undefined : ev.summary || undefined,
        };
        return {
          ...state,
          messages: [...msgs, { id: "res-" + ev.call_id, role: "assistant", content: "", toolCalls: [tc], streaming: false, createdAt: Date.now() }],
        };
      }
      return {
        ...state,
        messages: patchToolCall(msgs, ev.call_id, {
          phase: ev.ok ? "done" : "error",
          ok: ev.ok,
          summary: ev.summary || undefined,
          elapsedMs: ev.elapsed_ms, // 修"始终 0ms"(ToolResultMessage.elapsed_ms 实测值)
          attempts: ev.attempts,
          errorType: ev.ok ? undefined : ev.error_type || undefined,
          errorMessage: ev.ok ? undefined : ev.summary || undefined,
        }),
      };
    }

    case "TaskNotification": {
      // 后台子 agent 完成 -> 系统提示行(对齐 CLI 的 [后台任务] 打印;web/app 此前没有)
      // 幂等:同 run 的通知只插一条(restore 与 stream 重叠时防重复系统行)。
      const content = `[后台任务] ${ev.role} 完成(status=${ev.status})`;
      if (state.messages.some((m) => m.role === "system" && m.content === content)) return state;
      const notice: ChatMessage = {
        id: "tn-" + ev.run_id.slice(0, 6) + "-" + state.messages.length,
        role: "system",
        content,
        createdAt: Date.now(),
      };
      return { ...state, messages: [...state.messages, notice] };
    }

    case "RunEnd": {
      const finalText = ev.final_text || "";
      return {
        ...state,
        streaming: false,
        messages: state.messages.map((m, i) => {
          if (m.role !== "assistant") return m;
          const isLast = i === state.messages.length - 1;
          return {
            ...m,
            streaming: false,
            status: ev.status === "completed" ? "completed" : "failed",
            // 最后一条无文本(纯工具轮)时用 final_text 兜底
            content: m.content || (isLast ? finalText : ""),
            // 整轮总账挂最后一条 assistant(对齐 CC result:turn 结束才报 token 总统计)
            ...(isLast && ev.usage
              ? {
                  turnUsage: ev.usage,
                  durationMs: ev.duration_ms ?? undefined,
                  numSteps: ev.num_steps ?? undefined,
                }
              : {}),
          };
        }),
      };
    }

    default:
      return state;
  }
}
